### ISSUE TITLE
Adding temperature quantities can cause surprising results.

### ISSUE THREAD

**Issue by mikesamuel** [littledan/proposal-operator-overloading#6]
Temperature is not universally zeroed.

0°F ≠ 0°C ≠ 0K

but people expect that

*  0°C + 0°C = 0°C
*  0°F + 0°F = 0°F
*  0°K + 0°K = 0°K

If 0°C === 273.15°K then it would be confusing if 0°C + 0°K = 273.15°K.

In a previous life, I informally surveyed a bunch of mechanical engineers and they almost universally agreed that throwing out commutativity was preferable to throwing out the identity (x + 0\<unit\> = x) where x is numeric.

One way to think of this is to treat numeric addition as vector extension.
That way when the right operand is converted, you don't adjust for the different base, just for scale.

It would be nice if operator overloading allowed extended numeric types libraries that include temperatures to do this kind of late, scale-only coercion.

**Comment by littledan** [comment]
Sounds good. Is there anything in this proposal that would block this sort of usage, or are you just suggesting that we document this constraint in the README?

**Comment by littledan** [comment]
Yeah, based on the experience in other languages, I'm highly skeptical of adding a type promotion system to JavaScript.  Probably I should include the rationale for that in the README and also include a units example.

(Do people really want to add non-Kelvin temperatures? That seems totally wrong. I'd want to add a temperature to a temperature interval, I think.)

**Comment by mikesamuel** [comment]
> Do people really want to add non-Kelvin temperatures?

In my (limited, long ago) experience with scientific computation, people don't convert values from external sources unless the language requires it, so you might see something like

```js
const normalTemperature = 20 °C    // NTP per nist.gov at experiment start
const tempRate = 0.05 K / sec  // Just as likely to see 0.05 °C / sec
function temperatureAtTime(t) { return normalTemperature + tempRate * t; }
```

> That seems totally wrong. I'd want to add a temperature to a temperature interval, I think.)

Yeah, you can solve it by requiring a distinct type for the right operand when the left is a non-Kelvin temperature.

In systems I worked on, we did a lot of work to avoid surfacing that in the user-visible portion of the type system.

**Comment by littledan** [comment]
*frown emoji react!*

**Comment by littledan** [comment]
Thanks for the context!
